### PR TITLE
#3750 Updated force unique filter to work with references to separate topics inside the same file

### LINKED ISSUE #3750: Reconsider how force-unique works with references to separate topics inside the same file
I'm attaching a DITA Map which references a glossary.ditamap with the following content:

    <map>
        <glossref href="concepts/glossary.dita#flowers.genus" print="yes" keys="genus"/>
        <glossref href="concepts/glossary.dita#flowers.pollination" print="yes" keys="pollination"/>
      ....
    </map>

so we have multiple references to multiple glossentry elements located in the same file.
If I publish to PDF setting the "force-unique=true" parameter the PDF output will contain the glossary entries multiplied a number if times because the ForceUniqueFilter adds copy-to attributes to each reference.

I think the ForceUniqueFilter should not consider that references like "glossary.dita#flowers.genus" or "glossary.dita#flowers.pollination" are identical and thus avoid setting extra copy-to attributes on them.

[force-unique-flowers.zip](https://github.com/dita-ot/dita-ot/files/6458775/force-unique-flowers.zip)

### PR DESCRIPTION
**Pull request by BeniaminSavu**
Fix for #3750. Updated force unique filter to work with references to separate topics inside the same file.

Signed-off-by: Beniamin Savu <beniamin_savu@sync.ro>

## Description
Updated force unique filter to work with references to separate topics inside the same file.

## Motivation and Context
Fixes #3750

## How Has This Been Tested?
Manual testing using the samples from #3750 and with an automated test.

## Type of Changes
- Bug fix _(non-breaking change which fixes an issue)_

## Documentation and Compatibility
No docs update necessary.

## Checklist
- My code follows the code style of this project.
    -  <https://github.com/dita-ot/dita-ot/wiki/Java-Coding-Conventions>
    -  <https://github.com/dita-ot/dita-ot/wiki/XSLT-Coding-Conventions>
    -  <https://github.com/dita-ot/docs/wiki/coding-guidelines>
- I have updated the unit tests to reflect the changes in my code.

### CODE REVIEW

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java [hunk 1]
```diff
@@ -117,4 +151,46 @@ private FileInfo generateCopyToTarget(final FileInfo srcFi, final int count) {
...
+    
+    /**
+     * Holds additional information about the parent for a topicref.
+     *
+     */
+    private static final class ParentTopicref {
+      /**
+       * The href value for the parent topic.
+       */
+      private URI href;
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/writer/ForceUniqueFilter.java:163]: By coding convention, the fields here should be `public final` and getters should be removed

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java [hunk 2]
```diff
@@ -39,13 +40,26 @@
     public final Map<FileInfo, FileInfo> copyToMap = new HashMap<>();
     private TempFileNameScheme tempFileNameScheme;
 
+    /**
+     * Stack used to detect the current element type. 
+     */
+    private final Deque<String> classElementStack = new LinkedList<>();
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/writer/ForceUniqueFilter.java:46]: Could this also be `ArrayDeque`?

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java [hunk 3]
```diff
@@ -55,20 +69,35 @@ public void startElement(final String uri, final String localName, final String
...
                     (format == null || format.equals(ATTR_FORMAT_VALUE_DITA)) &&
                     (processingRole == null || processingRole.equals(ATTR_PROCESSING_ROLE_VALUE_NORMAL))) {
                 final URI file = stripFragment(source);
                 Integer count = topicrefCount.getOrDefault(file, 0);
-                count++;
-                topicrefCount.put(file, count);
+                
+                ParentTopicref parentTopicref = topicrefParentsStack.peek();
+                boolean parentTopicHasEmbededSubtopcis = false;
+                if(parentTopicref != null && parentTopicref.getHref() != null) {
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/writer/ForceUniqueFilter.java:82]: Add space between `if` and `(`, here and in other places

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java [hunk 4]
```diff
@@ -55,20 +69,35 @@ public void startElement(final String uri, final String localName, final String
...
+                  URI parentHref = parentTopicref.getHref();
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/writer/ForceUniqueFilter.java:83]: Indent should be 4 spaces

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java (hunk 1 above)
Comment (BeniaminSavu) [reply src/main/java/org/dita/dost/writer/ForceUniqueFilter.java]: I made the requested changes

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java (hunk 3 above)
Comment (BeniaminSavu) [reply src/main/java/org/dita/dost/writer/ForceUniqueFilter.java]: I made the requested changes

File: src/main/java/org/dita/dost/writer/ForceUniqueFilter.java (hunk 4 above)
Comment (BeniaminSavu) [reply src/main/java/org/dita/dost/writer/ForceUniqueFilter.java]: Fixed indentation

**Comment by jelovirt** [comment]
@BeniaminSavu could you rebase and add sign-off to commits, see https://github.com/dita-ot/dita-ot/pull/3894/checks?check_run_id=5389933154

**Comment by BeniaminSavu** [comment]
I rebased and added sign-off to commits

**Review by jelovirt — APPROVED** [review]

**Comment by chrispy-snps** [comment]
Thanks @BeniaminSavu, @jelovirt!

**Pull request merged**